internal/mocks/connection.go

```diff
@@ -0,0 +1,298 @@
+func (m *MockConnection) AppendReceiverAttach(linkName string, linkHandle uint32, mode encoding.ReceiverSettleMode) error {
+	return m.AppendFrame(FrameAMQP, &frames.PerformAttach{
+		Name:   linkName,
+		Handle: linkHandle,
+		Role:   encoding.RoleSender,
+		Source: &frames.Source{
+			Address:      "test",
+			Durable:      encoding.DurabilityNone,
+			ExpiryPolicy: encoding.ExpirySessionEnd,
+		},
+		ReceiverSettleMode: &mode,
+		MaxMessageSize:     math.MaxUint32,
+	})
+}
+
+// AppendTransfer appends a PerformTransfer frame with the specified values.
+// The linkHandle MUST match the linkHandle value specified in AppendReceiverAttach.
+func (m *MockConnection) AppendTransfer(linkHandle, deliveryID uint32, payload []byte) error {
+	format := uint32(0)
+	payloadBuf := &buffer.Buffer{}
+	encoding.WriteDescriptor(payloadBuf, encoding.TypeCodeApplicationData)
+	err := encoding.WriteBinary(payloadBuf, payload)
+	if err != nil {
+		return err
+	}
+	return m.AppendFrame(FrameAMQP, &frames.PerformTransfer{
+		Handle:        linkHandle,
+		DeliveryID:    &deliveryID,
+		DeliveryTag:   []byte("tag"),
+		MessageFormat: &format,
+		Payload:       payloadBuf.Detach(),
+	})
+}
+
+// AppendDisposition appends a PerformDisposition frame with the specified values.
+// The deliveryID MUST match the deliveryID value specified in AppendTransfer.
+func (m *MockConnection) AppendDisposition(deliveryID uint32, state encoding.DeliveryState) error {
+	return m.AppendFrame(FrameAMQP, &frames.PerformDisposition{
+		Role:    encoding.RoleSender,
+		First:   deliveryID,
+		Settled: true,
+		State:   state,
+	})
+}
+
+// AppendFrame adds the specified frame to the list of responses.
+func (m *MockConnection) AppendFrame(t FrameType, f frames.FrameBody) error {
+	raw, err := m.encodeFrame(t, f)
+	if err != nil {
+		return err
+	}
+	m.pendingDataMu.Lock()
+	defer m.pendingDataMu.Unlock()
+	m.pendingData = append(m.pendingData, raw)
+	return nil
+}
+
+func (m *MockConnection) encodeFrame(t FrameType, f frames.FrameBody) ([]byte, error) {
+	bodyBuf := buffer.New([]byte{})
+	if err := encoding.Marshal(bodyBuf, f); err != nil {
+		return nil, err
+	}
+	// create the frame header, needs size of the body plus itself
+	header := frameHeader{
+		Size:       uint32(bodyBuf.Len()) + 8,
+		DataOffset: 2,
+		FrameType:  uint8(t),
+	}
+	headerBuf := buffer.New([]byte{})
+	if err := encoding.Marshal(headerBuf, header); err != nil {
+		return nil, err
+	}
+	// concatenate header + body
+	raw := headerBuf.Detach()
+	raw = append(raw, bodyBuf.Detach()...)
+	return raw, nil
+}
+
+///////////////////////////////////////////////////////
+// following functions are for the net.Conn interface
+///////////////////////////////////////////////////////
+
+func (m *MockConnection) Read(b []byte) (n int, err error) {
+	if err := m.closedCheck(); err != nil {
+		return 0, err
+	}
+	select {
+	case <-m.readDL.C:
+		return 0, errors.New("read deadline exceeded")
```

Comment:
NIT: Would it be possible to return the actual error you'd expect to see here? `net.OpError` perhaps?

Or are all errors (Temporary or Timeout) all treated the same? @serbrech had some pretty nice error handling code in go-shuttle that's made me think about this more.

Reply:
At present, all errors are treated as terminal, see [here](https://github.com/Azure/go-amqp/blob/master/conn.go#L530-L547).  Might be a candidate for some retry, but that can be handled separately.

Generally speaking, trying to reason about network errors beyond nil/non-nil falls apart pretty fast.  You initially start with three buckets of error types.

- errors we know to be transient - retry
- errors we know aren't transient - don't retry
- errors we are unsure about - retry

This seems reasonable on the surface, but in practice, if you don't own the chain of calls end-to-end, you can't distinguish case 1 and case 2, so everything coalesces into case 3, so you just always end up retrying.  This is why we added the NonRetriableError interface and annotate errors we _know_ are non-transient.  The vast majority are considered non-transient because you just don't know.

Reply:
In goshuttle i progressively add error handling that identify when the connection inside the sdk is dead, and i reconstruct the publisher/listener in place. If I don't know the error yet, I let the service crash and restart. (A more brutal recover...)
That just forces me to be aware of the errors that do bubble out and I build up a robust in-process recovery layer over time.

Reply:
@serbrech , track 2 service bus borrows most of that code so similar constraints/benefits there.